src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplate.java

```diff
@@ -93,36 +95,32 @@ private AsyncDocumentClient getAsyncDocumentClient() {
     public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
     }
 
-    public <T> T insert(T objectToSave, PartitionKey partitionKey) {
-        Assert.notNull(objectToSave, "entityClass should not be null");
+    @Override
+    public <T> T insert(@NonNull String collectionName, @NonNull T domain, @Nullable PartitionKey key) {
+        Assert.hasText(collectionName, "collectionName should not be null, empty or only whitespaces");
+        Assert.notNull(domain, "domain should not be null");
 
-        return insert(getCollectionName(objectToSave.getClass()), objectToSave, partitionKey);
+        return insertAsync(collectionName, domain, key).toBlocking().single();
     }
 
-    public <T> T insert(String collectionName, T objectToSave, PartitionKey partitionKey) {
+    @Override
+    public <T> Observable<T> insertAsync(@NonNull String collectionName, @NonNull T domain,
+                                         @Nullable PartitionKey key) {
         Assert.hasText(collectionName, "collectionName should not be null, empty or only whitespaces");
-        Assert.notNull(objectToSave, "objectToSave should not be null");
-
-        final Document document = mappingDocumentDbConverter.writeDoc(objectToSave);
+        Assert.notNull(domain, "domain should not be null");
 
-        log.debug("execute createDocument in database {} collection {}", this.dbName, collectionName);
+        final String collectionLink = getCollectionLink(this.dbName, collectionName);
+        @SuppressWarnings("unchecked") final Class<T> domainClass = (Class<T>) domain.getClass();
+        final com.microsoft.azure.cosmosdb.Document document = mappingDocumentDbConverter.toCosmosdbDocument(domain);
 
-        try {
-            final Resource result = getDocumentClient()
-                    .createDocument(getCollectionLink(this.dbName, collectionName), document,
-                            getRequestOptions(partitionKey, null), false).getResource();
-
-            if (result instanceof Document) {
-                final Document documentInserted = (Document) result;
-                @SuppressWarnings("unchecked") final Class<T> domainClass = (Class<T>) objectToSave.getClass();
-
-                return mappingDocumentDbConverter.read(domainClass, documentInserted);
-            } else {
-                return null;
-            }
-        } catch (com.microsoft.azure.documentdb.DocumentClientException e) {
-            throw new DocumentDBAccessException("insert exception", e);
-        }
+        return getAsyncDocumentClient()
+                .createDocument(collectionLink, document, getRequestOptions(key, null), false)
+                .doOnNext(r -> log.info("Create Document Async from {}.", collectionLink))
```

Comment:
Will `info` log level  be too much as each insert will print the log, or use `debug`?

Reply:
make sense, should be in debug scope.